apps: Fix dc triggers reconciliation on image change and do not deploy DCs with empty image

tnozicka opened the pull request:
Fixes: https://github.com/openshift/origin/issues/17404
Fixes: https://github.com/openshift/origin/issues/17535
Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=1515902

/cc @mfojtik

mfojtik commented on pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -160,6 +160,15 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
+
+	// Never deploy with invalid or unresolved images
+	for i, container := range config.Spec.Template.Spec.Containers {
+		if strings.TrimSpace(container.Image) == "" {
```
i thought we guard against this in `shouldTrigger()`

tnozicka replied:
we don't, we wait for the RC to be created and because the image resolution/injection is happening asynchronously it will manage to deploy RC with "" before the image gets injected

tnozicka replied:
I tried fixing the image that function but then it will end up being triggered as config change. we could probably add exception for that as well but this felt cleanest and most reliable. We should do the same upstream.

mfojtik replied:
would it be better to move this up to `triggerActivated` and just return false, true ?

tnozicka replied:
I guess it could be there now that shouldSkip is honored

tnozicka replied:
@mfojtik actually that wouldn't work for manual rollouts without any trigger, so need to stay here

tnozicka commented:
/retest

tnozicka commented:
Added a fix for DC image reactor. Tests are green and ready for review.
/assign @mfojtik

tnozicka commented:
/cherrypick release-3.8
/cherrypick release-3.7

tnozicka commented:
/cherrypick release-3.7

mfojtik commented on pkg/image/trigger/deploymentconfigs/deploymentconfigs.go:
```diff
@@ -128,13 +128,19 @@ func (i deploymentConfigTriggerIndexer) Index(obj, old interface{}) (string, *tr
+		// We need to react on image changes as well. Image names could change,
+		// images could be set to different value or resetted to "" e.g. by oc apply
+		// and we need to make sure those changes get reconciled by re-resolving images
+		case !reflect.DeepEqual(dc.Spec.Template.Spec.Containers, oldDC.Spec.Template.Spec.Containers):
```
should be check only the images then?

mfojtik replied:
worst case this will be processed but not triggered, but I think we need some common helper like HasUpdatedImages() to work well for this and for other places. this is fine now.

tnozicka replied:
discussed on IRC but for the record: I don't see other way we can determine if it needs to be queued without actually doing all the work in UpdateDeploymentConfigImages. This is just a check for adding to work queue, it should be fast and simple. After picking it up from queue we already do all the checks for updated images and others.

kargakis replied:
What happens when I want to run my own image for sometime and not the image resolved by the trigger? Can I do that now?

tnozicka replied:
@kargakis set `automatic` to false
if it's active the state should always be reconciled

mfojtik commented:
one last question, looks good to me otherwise

mfojtik commented:
/lgtm

tnozicka commented:
/retest

tnozicka commented:
/retest

tnozicka commented:
/retest

tnozicka commented:
rebased the only change is 
```
interdiff <(git diff origin/fix-dc-triggers-when-patching-empty-image~3..origin/fix-dc-triggers-when-patching-empty-image) <(git diff upstream..fix-dc-triggers-when-patching-empty-image) | colordiff 
reverted:
--- b/test/extended/deployments/util.go
+++ a/test/extended/deployments/util.go
@@ -30,6 +30,7 @@
 
 	deployapiv1 "github.com/openshift/api/apps/v1"
 	deployapi "github.com/openshift/origin/pkg/apps/apis/apps"
+	deployconversionsv1 "github.com/openshift/origin/pkg/apps/apis/apps/v1"
 	appstypedclientset "github.com/openshift/origin/pkg/apps/generated/internalclientset/typed/apps/internalversion"
 	deployutil "github.com/openshift/origin/pkg/apps/util"
 	exutil "github.com/openshift/origin/test/extended/util"
@@ -633,7 +634,7 @@
 	}
 
 	dc := new(deployapi.DeploymentConfig)
+	err = deployconversionsv1.Convert_v1_DeploymentConfig_To_apps_DeploymentConfig(dcv1, dc, nil)
-	err = kapi.Scheme.Convert(dcv1, dc, nil)
 	return dc, err
 }
```

tnozicka commented:
/retest

tnozicka commented:
/retest

tnozicka commented:
/retest

tnozicka commented:
@mfojtik this needs to be re-lgtm, thx

mfojtik commented:
/lgtm

(first lgtm this year \o/ ;-)

tnozicka commented:
/retest

tnozicka commented:
/hold

there seems to be a race in DC sometimes creating RC2 sometimes not resulting in flakes; figuring out why

tnozicka commented:
/hold cancel

tnozicka commented on pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -128,28 +128,44 @@ func (c *DeploymentConfigController) Handle(config *appsapi.DeploymentConfig) er
-	shouldTrigger, shouldSkip := triggerActivated(configCopy, latestExists, latestDeployment, c.codec)
-	if !shouldSkip && shouldTrigger {
-		configCopy.Status.LatestVersion++
-		return c.updateStatus(configCopy, existingDeployments)
```
@mfojtik I wonder how this ever worked
https://github.com/openshift/origin/blob/69d5e205544faf41e433f3af395dccf2864fe163/pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go#L324-L326

it doesn't see `latestVersion` change (made prior to calculating status) and if there is no other change the update gets ignored

mfojtik replied:
¯\_(ツ)_/¯

tnozicka commented:
/test cmd

mfojtik commented on pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -128,28 +128,44 @@ func (c *DeploymentConfigController) Handle(config *appsapi.DeploymentConfig) er
+
+	// Never deploy with invalid or unresolved images
+	for i, container := range config.Spec.Template.Spec.Containers {
+		if strings.TrimSpace(container.Image) == "" {
```
len() == 0

tnozicka replied:
I've noticed you like it :) I tend to write `== ""` as it say more what is meant here. compiler should optimize it away

mfojtik replied:
it is a consistency thing, I used to use `== ""` as well, but @smarterclayton always pointed it out in reviews so I learned to use len() ;-)

tnozicka replied:
harder to read, but as you wish ;)

tnozicka replied:
done

mfojtik commented on pkg/image/trigger/deploymentconfigs/deploymentconfigs.go:
```diff
@@ -187,9 +193,6 @@ func UpdateDeploymentConfigImages(dc *appsapi.DeploymentConfig, tagRetriever tri
 			glog.V(4).Infof("trigger %#v in deployment %s is not resolveable", p, dc.Name)
 			return nil, false, nil
 		}
-		if ref == p.LastTriggeredImage {
```
why is this going away?

tnozicka replied:
because it will stop DC...image from being reconciled. If someone overwrites it manually, this will prevent reconciliation which should have been done

mfojtik commented on pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -128,28 +128,44 @@ func (c *DeploymentConfigController) Handle(config *appsapi.DeploymentConfig) er
 		// want to compete with the deployer.
 		if !appsutil.IsTerminatedDeployment(latestDeployment) {
-			return c.updateStatus(config, existingDeployments)
+			return c.updateStatus(config, existingDeployments, false)
```
is this the only case when we don't update the observed generation? maybe it will be nicer to have two funcs: `updateStatusAndGeneration` and `updateStatusWithoutGeneration`... (i really don't like the boolean params as one have to always search what are they for ;-)

tnozicka replied:
I don't think it's worth it to split 1 func into 3 (1 internal) just because of the bool; I don't feel strong about it though

mfojtik commented on pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -394,10 +419,17 @@ func calculateStatus(config *appsapi.DeploymentConfig, rcs []*v1.ReplicationCont
 		unavailableReplicas = 0
 	}
 
+	var generation int64
```
get rid of "else" with:

```go
generation := config.Status.ObservedGeneration
if updateObservedGeneration {
  generation = config.Generation
}
```

tnozicka replied:
getting rid of else would mean two writes instead of one, but I can change it if you want

mfojtik replied:
write is fine, else is harder to brain scan :)

tnozicka replied:
done

mfojtik commented on pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -542,50 +574,49 @@ func triggerActivated(config *appsapi.DeploymentConfig, latestExists bool, lates
 		if err != nil {
-			glog.Errorf("Error while checking for latest pod template in replication controller: %v", err)
-			return false, true
+			return false, false, fmt.Errorf("error while checking for latest pod template in replication controller: %v", err)
```
why returning this error? isn't logging it enough?

tnozicka replied:
if you return error it will get rate limited if it's repeating and it should not update status as it failed to act on it

mfojtik replied:
appsutil.HasLatestPodTemplate can only fail when decoding deployment config, I guess it won't even get this far if decoding fail...

Maybe we can convert the second bool to an error and then check that error time (ErrSkipTrigger) or something?

tnozicka replied:
> appsutil.HasLatestPodTemplate can only fail when decoding deployment config, I guess it won't even get this far if decoding fail...

which is likely a permanent failure which the ratelimiting queue will shield us from

> Maybe we can convert the second bool to an error and then check that error time (ErrSkipTrigger) or something?

seems too obscure; skipping the trigger is not an error

but if it fails and gets here, the queue rate limitting needs to kick in because i

tnozicka commented:
/test cmd

tnozicka commented:
/retest

tnozicka commented:
```diff
diff --git a/pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go b/pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go
index 5066921448..cd882c4d28 100644
--- a/pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go
+++ b/pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go
@@ -131,7 +131,7 @@ func (c *DeploymentConfigController) Handle(config *appsapi.DeploymentConfig) er
 
        // Never deploy with invalid or unresolved images
        for i, container := range config.Spec.Template.Spec.Containers {
-               if strings.TrimSpace(container.Image) == "" {
+               if len(strings.TrimSpace(container.Image)) == 0 {
                        glog.V(4).Infof("Postponing rollout #%d for DeploymentConfig %s/%s because of invalid or unresolved image for container #%d (name=%s)", config.Status.LatestVersion, config.Namespace, config.Name, i, container.Name)
                        return c.updateStatus(config, existingDeployments, true)
                }
@@ -419,10 +419,8 @@ func calculateStatus(config *appsapi.DeploymentConfig, rcs []*v1.ReplicationCont
                unavailableReplicas = 0
        }
 
-       var generation int64
+       generation := config.Generation
        if updateObservedGeneration {
-               generation = config.Generation
-       } else {
                generation = config.Status.ObservedGeneration
        }
```

tnozicka commented:
```diff
diff --git a/pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go b/pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go
index 5066921448..cd882c4d28 100644
--- a/pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go
+++ b/pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go
@@ -131,7 +131,7 @@ func (c *DeploymentConfigController) Handle(config *appsapi.DeploymentConfig) er
 
        // Never deploy with invalid or unresolved images
        for i, container := range config.Spec.Template.Spec.Containers {
-               if strings.TrimSpace(container.Image) == "" {
+               if len(strings.TrimSpace(container.Image)) == 0 {
                        glog.V(4).Infof("Postponing rollout #%d for DeploymentConfig %s/%s because of invalid or unresolved image for container #%d (name=%s)", config.Status.LatestVersion, config.Namespace, config.Name, i, container.Name)
                        return c.updateStatus(config, existingDeployments, true)
                }
@@ -419,10 +419,8 @@ func calculateStatus(config *appsapi.DeploymentConfig, rcs []*v1.ReplicationCont
                unavailableReplicas = 0
        }
 
-       var generation int64
+       generation := config.Status.ObservedGeneration
        if updateObservedGeneration {
-               generation = config.Generation
-       } else {
                generation = config.Generation
        }
```

tnozicka commented:
/hold

mfojtik commented:
/lgtm now, thanks!

tnozicka commented:
hmm, somehow this flakes:
```
/go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/test/extended/deployments/deployments.go:1390
Expected error:
    <*errors.errorString | 0xc4201ede90>: {
        s: "timed out waiting for the condition",
    }
    timed out waiting for the condition
not to have occurred
/go/src/github.com/openshift/origin/_output/local/go/src/github.com/openshift/origin/test/extended/deployments/deployments.go:1447
```
run this 50 times on my machine and no flake
maybe it was overloaded and didn't make it in 30s; adding test logs to see

tnozicka commented:
other flakes
/retest

tnozicka commented:
github down on clone
/retest

tnozicka commented:
/retest

tnozicka commented:
/retest

tnozicka commented:
no changes, just squashed the commit with additional log message for tests

/hold cancel
we should be good to go and ready with logs if we see that flake again

tnozicka commented:
/retest

tnozicka commented:
infra issues
/retest

tnozicka commented:
```diff
diff --git a/pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go b/pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go
index 4ae6a2a202..e309e572a0 100644
--- a/pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go
+++ b/pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go
@@ -121,6 +121,17 @@ func (c *DeploymentConfigController) Handle(config *appsapi.DeploymentConfig) er
 		return c.updateStatus(config, existingDeployments, true)
 	}
 
+	// If the config is paused we shouldn't create new deployments for it.
+	if config.Spec.Paused {
+		// in order for revision history limit cleanup to work for paused
+		// deployments, we need to trigger it here
+		if err := c.cleanupOldDeployments(existingDeployments, config); err != nil {
+			c.recorder.Eventf(config, v1.EventTypeWarning, "DeploymentCleanupFailed", "Couldn't clean up deployments: %v", err)
+		}
+
+		return c.updateStatus(config, existingDeployments, true)
+	}
+
 	latestExists, latestDeployment := appsutil.LatestDeploymentInfo(config, existingDeployments)
 
 	if !latestExists {
@@ -166,17 +177,6 @@ func (c *DeploymentConfigController) Handle(config *appsapi.DeploymentConfig) er
 		return c.reconcileDeployments(existingDeployments, config, cm)
 	}
 
-	// If the config is paused we shouldn't create new deployments for it.
-	if config.Spec.Paused {
-		// in order for revision history limit cleanup to work for paused
-		// deployments, we need to trigger it here
-		if err := c.cleanupOldDeployments(existingDeployments, config); err != nil {
-			c.recorder.Eventf(config, v1.EventTypeWarning, "DeploymentCleanupFailed", "Couldn't clean up deployments: %v", err)
-		}
-
-		return c.updateStatus(config, existingDeployments, true)
-	}
-
 	// Never deploy with invalid or unresolved images
 	for i, container := range config.Spec.Template.Spec.Containers {
 		if strings.TrimSpace(container.Image) == "" {
diff --git a/test/extended/deployments/deployments.go b/test/extended/deployments/deployments.go
index 580cfc1d0b..593e456658 100644
--- a/test/extended/deployments/deployments.go
+++ b/test/extended/deployments/deployments.go
@@ -14,7 +14,6 @@ import (
 	kapiv1 "k8s.io/api/core/v1"
 	kerrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/fields"
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/apimachinery/pkg/watch"
@@ -701,13 +700,15 @@ var _ = g.Describe("[Feature:DeploymentConfig] deploymentconfigs", func() {
 	})
 
 	g.Describe("paused [Conformance]", func() {
+		dcName := "paused"
 		g.AfterEach(func() {
-			failureTrap(oc, "paused", g.CurrentGinkgoTestDescription().Failed)
+			failureTrap(oc, dcName, g.CurrentGinkgoTestDescription().Failed)
 		})
 
 		g.It("should disable actions on deployments", func() {
 			resource, name, err := createFixture(oc, pausedDeploymentFixture)
 			o.Expect(err).NotTo(o.HaveOccurred())
+			o.Expect(name).To(o.Equal(dcName))
 
 			_, rcs, _, err := deploymentInfo(oc, name)
 			o.Expect(err).NotTo(o.HaveOccurred())
@@ -757,6 +758,20 @@ var _ = g.Describe("[Feature:DeploymentConfig] deploymentconfigs", func() {
 			})
 			o.Expect(err).NotTo(o.HaveOccurred())
 			o.Expect(waitForLatestCondition(oc, name, deploymentRunTimeout, deploymentReachedCompletion)).NotTo(o.HaveOccurred())
+
+			g.By("making sure it updates observedGeneration after being paused")
+			dc, err := oc.AppsClient().Apps().DeploymentConfigs(oc.Namespace()).Patch(dcName,
+				types.StrategicMergePatchType, []byte(`{"spec": {"paused": true}}`))
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			_, err = waitForDCModification(oc, dc.Namespace, dc.Name, deploymentChangeTimeout,
+				dc.GetResourceVersion(), func(config *appsapi.DeploymentConfig) (bool, error) {
+					if config.Status.ObservedGeneration >= dc.Generation {
+						return true, nil
+					}
+					return false, nil
+				})
+			o.Expect(err).NotTo(o.HaveOccurred(), fmt.Sprintf("failed to wait on generation >= %d to be observed by DC %s/%s", dc.Generation, dc.Namespace, dc.Name))
 		})
 	})
 
@@ -1395,10 +1410,7 @@ var _ = g.Describe("[Feature:DeploymentConfig] deploymentconfigs", func() {
 			o.Expect(err).NotTo(o.HaveOccurred())
 			o.Expect(dc.Name).To(o.Equal(dcName))
 
-			watcher, err := oc.KubeClient().CoreV1().ReplicationControllers(namespace).Watch(
-				metav1.ListOptions{
-					FieldSelector: fields.OneTermEqualSelector("metadata.name", rcName(1)).String(),
-				})
+			rcList, err := oc.KubeClient().CoreV1().ReplicationControllers(namespace).List(metav1.ListOptions{})
 			o.Expect(err).NotTo(o.HaveOccurred())
 
 			dc.Spec.Replicas = 1
@@ -1410,13 +1422,18 @@ var _ = g.Describe("[Feature:DeploymentConfig] deploymentconfigs", func() {
 			e2e.Logf("%s", out)
 			o.Expect(err).NotTo(o.HaveOccurred())
 
-			g.By("waiting for RC #1 to be created")
-			_, err = watch.Until(deploymentChangeTimeout, watcher, func(event watch.Event) (bool, error) {
-				if event.Type == watch.Added {
-					return true, nil
-				}
-				return false, fmt.Errorf("different kind of event appeared while waiting for Added event: %#v", event)
-			})
+			g.By("waiting for deployment #1 to complete")
+			_, err = waitForRCModification(oc, namespace, rcName(1), deploymentRunTimeout,
+				rcList.ResourceVersion, func(currentRC *kapiv1.ReplicationController) (bool, error) {
+					switch appsutil.DeploymentStatusFor(currentRC) {
+					case appsapi.DeploymentStatusComplete:
+						return true, nil
+					case appsapi.DeploymentStatusFailed:
+						return true, fmt.Errorf("deployment #1 failed")
+					default:
+						return false, nil
+					}
+				})
 			o.Expect(err).NotTo(o.HaveOccurred())
 
 			g.By("setting DC image repeatedly to empty string to fight with image trigger")
```

tnozicka commented:
@mfojtik [this is](https://github.com/openshift/origin/blob/56ed4ecfe9f296e2f31ff90171a79037091e8846/pkg/apps/cmd/delete.go#L63) what was slowing down the **ci/openshift-jenkins/cmd** tests and timing out, should be fixed now

mfojtik commented on pkg/apps/controller/deploymentconfig/deploymentconfig_controller.go:
```diff
@@ -128,38 +139,52 @@ func (c *DeploymentConfigController) Handle(config *appsapi.DeploymentConfig) er
-		return c.updateStatus(config, existingDeployments)
+	// Never deploy with invalid or unresolved images
+	for i, container := range config.Spec.Template.Spec.Containers {
+		if strings.TrimSpace(container.Image) == "" {
```
len() :-)

tnozicka replied:
fixed, hopefully for good now

tnozicka commented:
CI cluster network issues
/retest

tnozicka commented:
flake: https://github.com/openshift/origin/issues/17882
/retest

tnozicka commented:
@mfojtik len() is there, tests are green - lgty?

mfojtik commented:
/lgtm

tnozicka commented:
/cherrypick release-3.8
